File: menu-flutter/lib/services/review_service.dart
import '../core/api_client.dart';

/// 点评服务（V43：食集评价 + 我的评价 + 单菜均分）。
class ReviewService {
  /// 点评维度字典：GET /dict?group=review_dimension
  static Future<List<ReviewDimension>> dimensions() async {
    final data = await ApiClient.instance.get(
      '/dict',
      query: {'group': 'review_dimension'},
    );
    if (data is List) {
      return data
          .map((e) => ReviewDimension.fromJson(e as Map<String, dynamic>))
          .toList();
    }
    if (data is Map && data['records'] is List) {
      return (data['records'] as List)
          .map((e) => ReviewDimension.fromJson(e as Map<String, dynamic>))
          .toList();
    }
    return [];
  }

  /// 提交点评（dishId/menuId 二选一）：POST /review
  static Future<void> submitReview(Map<String, dynamic> data) async {
    await ApiClient.instance.post('/review', body: data);
  }

  /// 统一评价页数据：GET /review/menu-overview/{menuId}
  static Future<MenuReviewOverview> menuOverview(int menuId) async {
    final data = await ApiClient.instance.get('/review/menu-overview/$menuId');
    return MenuReviewOverview.fromJson(data as Map<String, dynamic>);
  }

  /// 我的评价：GET /review/mine
  static Future<MyReviews> myReviews() async {
    final data = await ApiClient.instance.get('/review/mine');
    return MyReviews.fromJson(data as Map<String, dynamic>);
  }

  /// 单菜均分 + 评价数（菜谱详情展示）：GET /review/dish/{dishId}/avg
  static Future<(double?, int)> dishAvg(int dishId) async {
    final data = await ApiClient.instance.get('/review/dish/$dishId/avg');
    if (data is Map && data['star'] is String) {
      return (double.tryParse(data['star'] as String), (data['count'] as num?)?.toInt() ?? 0);
    }
    return (null, 0);
  }
}

/// 点评维度（如"味道""口感""外观"）。
class ReviewDimension {
  final int id;
  final String name;

  const ReviewDimension({required this.id, required this.name});

  factory ReviewDimension.fromJson(Map<String, dynamic> j) => ReviewDimension(
        id: (j['id'] as num).toInt(),
        name: (j['name'] ?? '') as String,
      );
}

/// 统一评价页数据（GET /review/menu-overview/{menuId}）。
class MenuReviewOverview {
  final int menuId;
  final String menuName;
  final String? finishedAt;
  final int dishCount;
  final MenuReviewStatus? menuReview;
  final List<DishReviewStatus> dishes;

  const MenuReviewOverview({
    required this.menuId,
    required this.menuName,
    this.finishedAt,
    required this.dishCount,
    this.menuReview,
    required this.dishes,
  });

  factory MenuReviewOverview.fromJson(Map<String, dynamic> j) =>
      MenuReviewOverview(
        menuId: (j['menuId'] as num).toInt(),
        menuName: (j['menuName'] ?? '') as String,
        finishedAt: j['finishedAt'] as String?,
        dishCount: (j['dishCount'] as num?)?.toInt() ?? 0,
        menuReview: j['menuReview'] == null
            ? null
            : MenuReviewStatus.fromJson(j['menuReview'] as Map<String, dynamic>),
        dishes: ((j['dishes'] as List?) ?? [])
            .map((e) => DishReviewStatus.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

/// 食集整体评价状态（我的最近一条）。
class MenuReviewStatus {
  final bool reviewed;
  final int? starRating;
  final Map<int, int> dimensionScores;
  final String? createTime;

  const MenuReviewStatus({
    required this.reviewed,
    this.starRating,
    required this.dimensionScores,
    this.createTime,
  });

  factory MenuReviewStatus.fromJson(Map<String, dynamic> j) => MenuReviewStatus(
        reviewed: (j['reviewed'] as bool?) ?? false,
        starRating: (j['starRating'] as num?)?.toInt(),
        dimensionScores: ((j['dimensionScores'] as Map?) ?? {})
            .map((k, v) => MapEntry(int.parse(k as String), (v as num).toInt())),
        createTime: j['createTime'] as String?,
      );
}

/// 单道菜评价状态（我的最近一条）。
class DishReviewStatus {
  final int dishId;
  final String dishName;
  final String? coverUrl;
  final int? starRating;

  const DishReviewStatus({
    required this.dishId,
    required this.dishName,
    this.coverUrl,
    this.starRating,
  });

  factory DishReviewStatus.fromJson(Map<String, dynamic> j) => DishReviewStatus(
        dishId: (j['dishId'] as num).toInt(),
        dishName: (j['dishName'] ?? '') as String,
        coverUrl: j['coverUrl'] as String?,
        starRating: (j['starRating'] as num?)?.toInt(),
      );
}

/// 我的评价（GET /review/mine）。
class MyReviews {
  final List<ReviewEntry> reviews;
  final List<PendingMenu> pendingMenus;

  const MyReviews({required this.reviews, required this.pendingMenus});

  factory MyReviews.fromJson(Map<String, dynamic> j) => MyReviews(
        reviews: ((j['reviews'] as List?) ?? [])
            .map((e) => ReviewEntry.fromJson(e as Map<String, dynamic>))
            .toList(),
        pendingMenus: ((j['pendingMenus'] as List?) ?? [])
            .map((e) => PendingMenu.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

/// 评价历史条目（食集或菜品）。
class ReviewEntry {
  final int id;
  final int? dishId;
  final int? menuId;
  final String? name;
  final int starRating;
  final String? createTime;

  const ReviewEntry({
    required this.id,
    this.dishId,
    this.menuId,
    this.name,
    required this.starRating,
    this.createTime,
  });

  bool get isMenu => dishId == null && menuId != null;

  factory ReviewEntry.fromJson(Map<String, dynamic> j) => ReviewEntry(
        id: (j['id'] as num).toInt(),
        dishId: (j['dishId'] as num?)?.toInt(),
        menuId: (j['menuId'] as num?)?.toInt(),
        name: j['name'] as String?,
        starRating: (j['starRating'] as num).toInt(),
        createTime: j['createTime'] as String?,
      );
}

/// 待评价食集。
class PendingMenu {
  final int menuId;
  final String menuName;
  final String? finishedAt;
  final int dishCount;
  final int reviewedDishCount;
  final bool menuReviewed;

  const PendingMenu({
    required this.menuId,
    required this.menuName,
    this.finishedAt,
    required this.dishCount,
    required this.reviewedDishCount,
    required this.menuReviewed,
  });

  /// 还差几道菜没评（食集整体没评也算一道）。
  int get remaining {
    var r = dishCount - reviewedDishCount;
    if (!menuReviewed) r += 1;
    return r;
  }

  factory PendingMenu.fromJson(Map<String, dynamic> j) => PendingMenu(
        menuId: (j['menuId'] as num).toInt(),
        menuName: (j['menuName'] ?? '') as String,
        finishedAt: j['finishedAt'] as String?,
        dishCount: (j['dishCount'] as num?)?.toInt() ?? 0,
        reviewedDishCount: (j['reviewedDishCount'] as num?)?.toInt() ?? 0,
        menuReviewed: (j['menuReviewed'] as bool?) ?? false,
      );
}

File: menu-api/src/main/java/com/gudu/xsd/modules/review/ReviewService.java
package com.gudu.xsd.modules.review;

import cn.dev33.satoken.stp.StpUtil;
import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import com.gudu.xsd.common.BizException;
import com.gudu.xsd.modules.dish.Dish;
import com.gudu.xsd.modules.dish.mapper.DishMapper;
import com.gudu.xsd.modules.menu.Menu;
import com.gudu.xsd.modules.menu.MenuDish;
import com.gudu.xsd.modules.menu.mapper.MenuDishMapper;
import com.gudu.xsd.modules.menu.mapper.MenuMapper;
import com.gudu.xsd.modules.review.mapper.ReviewMapper;
import com.gudu.xsd.modules.review.mapper.ReviewScoreMapper;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * 点评服务（V43：支持食集评价）。
 *
 * <p>食集整体评价复用单菜四维度（review_score），review 表 dish_id/menu_id 二选一；
 * 单菜评价不关联食集（纯菜品维度）。
 *
 * <p>已评状态按「当前就餐成员」判定（评价记录 member 维度，食集本身不归属成员）。
 */
@Service
public class ReviewService {

    static final String MENU_STATUS_DONE = "DONE";

    private final ReviewMapper reviewMapper;
    private final ReviewScoreMapper reviewScoreMapper;
    private final MenuMapper menuMapper;
    private final MenuDishMapper menuDishMapper;
    private final DishMapper dishMapper;

    public ReviewService(ReviewMapper reviewMapper, ReviewScoreMapper reviewScoreMapper,
                         MenuMapper menuMapper, MenuDishMapper menuDishMapper, DishMapper dishMapper) {
        this.reviewMapper = reviewMapper;
        this.reviewScoreMapper = reviewScoreMapper;
        this.menuMapper = menuMapper;
        this.menuDishMapper = menuDishMapper;
        this.dishMapper = dishMapper;
    }

    /** 提交点评（菜品或食集）：当前就餐成员 + 级联维度分。 */
    @Transactional
    public Long submit(ReviewSaveDTO dto) {
        if ((dto.getDishId() == null) == (dto.getMenuId() == null)) {
            throw new BizException("菜品和食集评价二选一");
        }
        Long memberId = StpUtil.getSession().getLong("currentMemberId");
        if (memberId == null) throw new BizException("请先选择就餐成员");
        Review r = new Review();
        r.setDishId(dto.getDishId());
        r.setMenuId(dto.getMenuId());
        r.setMemberId(memberId);
        r.setStarRating(dto.getStarRating());
        r.setText(dto.getText());
        r.setImages(dto.getImages() == null ? null : String.join(",", dto.getImages()));
        reviewMapper.insert(r);
        if (dto.getDimensionScores() != null) {
            dto.getDimensionScores().forEach((dimId, score) -> {
                if (score == null || score < 1 || score > 5) throw new BizException("维度分需为 1-5");
                ReviewScore s = new ReviewScore();
                s.setReviewId(r.getId());
                s.setDimensionId(dimId);
                s.setScore(score);
                reviewScoreMapper.insert(s);
            });
        }
        return r.getId();
    }

    /** 某菜的所有点评（最新优先）。 */
    public List<Review> listByDish(Long dishId) {
        return reviewMapper.selectList(
            new QueryWrapper<Review>().eq("dish_id", dishId).orderByDesc("create_time"));
    }

    /** 某食集的所有整体评价（最新优先）。V43。 */
    public List<Review> listByMenu(Long menuId) {
        return reviewMapper.selectList(
            new QueryWrapper<Review>().eq("menu_id", menuId).orderByDesc("create_time"));
    }

    /** 统一评价页数据：食集信息 + 食集整体已评状态（我的）+ 每道菜已评状态（我的）。V43。 */
    public MenuReviewOverviewVO menuOverview(Long menuId) {
        Menu menu = menuMapper.selectById(menuId);
        if (menu == null) throw new BizException("食集不存在");
        List<Long> dishIds = menuDishMapper.selectList(new QueryWrapper<MenuDish>().eq("menu_id", menuId))
                .stream().map(MenuDish::getDishId).filter(Objects::nonNull).distinct().toList();
        Map<Long, Dish> dishMap = dishIds.isEmpty() ? Map.of()
                : dishMapper.selectBatchIds(dishIds).stream()
                        .collect(Collectors.toMap(Dish::getId, Function.identity(), (a, b) -> a));
        Long memberId = currentMemberId();

        // 食集整体评价（我的最近一条）
        MenuReviewOverviewVO.MenuReviewStatus menuReview = null;
        if (memberId != null) {
            Review mine = latestOne(reviewMapper.selectList(new QueryWrapper<Review>()
                    .eq("menu_id", menuId).eq("member_id", memberId).orderByDesc("create_time")));
            if (mine != null) {
                Map<Long, Integer> dims = reviewScoreMapper.selectList(
                                new QueryWrapper<ReviewScore>().eq("review_id", mine.getId())).stream()
                        .collect(Collectors.toMap(ReviewScore::getDimensionId, ReviewScore::getScore, (a, b) -> a));
                menuReview = new MenuReviewOverviewVO.MenuReviewStatus(true, mine.getStarRating(),
                        dims, mine.getCreateTime());
            }
        }
        // 每道菜已评状态（我的最近一条）
        Map<Long, Integer> starByDish = new LinkedHashMap<>();
        if (memberId != null && !dishIds.isEmpty()) {
            reviewMapper.selectList(new QueryWrapper<Review>()
                            .in("dish_id", dishIds).eq("member_id", memberId).orderByDesc("create_time"))
                    .forEach(r -> starByDish.putIfAbsent(r.getDishId(), r.getStarRating()));
        }
        List<MenuReviewOverviewVO.DishReviewStatus> dishes = new ArrayList<>();
        for (Long dishId : dishIds) {
            Dish d = dishMap.get(dishId);
            if (d == null) continue;
            dishes.add(new MenuReviewOverviewVO.DishReviewStatus(
                    dishId, d.getName(), d.getCoverUrl(), starByDish.get(dishId)));
        }
        return new MenuReviewOverviewVO(menuId, menu.getName(), menu.getFinishedAt(),
                dishes.size(), menuReview, dishes);
    }

    /** 我的评价：评价历史（食集+菜品）+ 待评价食集（已完成且没评完）。V43。 */
    public MyReviewsVO mine() {
        Long memberId = currentMemberId();
        if (memberId == null) {
            return new MyReviewsVO(List.of(), List.of());
        }
        // 评价历史
        List<Review> reviews = reviewMapper.selectList(new QueryWrapper<Review>()
                .eq("member_id", memberId).orderByDesc("create_time"));
        Set<Long> dishIds = reviews.stream().map(Review::getDishId).filter(Objects::nonNull).collect(Collectors.toSet());
        Set<Long> menuIds = reviews.stream().map(Review::getMenuId).filter(Objects::nonNull).collect(Collectors.toSet());
        Map<Long, String> dishName = dishIds.isEmpty() ? Map.of()
                : dishMapper.selectBatchIds(dishIds).stream()
                        .collect(Collectors.toMap(Dish::getId, Dish::getName, (a, b) -> a));
        Map<Long, String> menuName = menuIds.isEmpty() ? Map.of()
                : menuMapper.selectBatchIds(menuIds).stream()
                        .collect(Collectors.toMap(Menu::getId, Menu::getName, (a, b) -> a));
        List<MyReviewsVO.ReviewEntry> entries = reviews.stream()
                .map(r -> new MyReviewsVO.ReviewEntry(
                        r.getId(), r.getDishId(), r.getMenuId(),
                        r.getDishId() != null ? dishName.get(r.getDishId()) : menuName.get(r.getMenuId()),
                        r.getStarRating(), r.getCreateTime()))
                .filter(e -> e.name() != null)
                .toList();

        // 待评价食集：DONE 且（食集整体没评 或 还有菜没评）
        List<Menu> doneMenus = menuMapper.selectList(new QueryWrapper<Menu>()
                .eq("status", MENU_STATUS_DONE).orderByDesc("finished_at"));
        List<MyReviewsVO.PendingMenu> pending = new ArrayList<>();
        if (!doneMenus.isEmpty()) {
            List<Long> doneMenuIds = doneMenus.stream().map(Menu::getId).toList();
            List<MenuDish> allMds = menuDishMapper.selectList(
                    new QueryWrapper<MenuDish>().in("menu_id", doneMenuIds));
            Map<Long, List<MenuDish>> mdsByMenu = allMds.stream()
                    .collect(Collectors.groupingBy(MenuDish::getMenuId));
            Set<Long> myReviewedMenus = reviewMapper.selectList(new QueryWrapper<Review>()
                            .eq("member_id", memberId).in("menu_id", doneMenuIds)).stream()
                    .map(Review::getMenuId).collect(Collectors.toSet());
            Set<Long> allDishIds = allMds.stream().map(MenuDish::getDishId)
                    .filter(Objects::nonNull).collect(Collectors.toSet());
            Set<Long> myReviewedDishes = allDishIds.isEmpty() ? Set.of()
                    : reviewMapper.selectList(new QueryWrapper<Review>()
                            .eq("member_id", memberId).in("dish_id", allDishIds)).stream()
                            .map(Review::getDishId).collect(Collectors.toSet());
            for (Menu m : doneMenus) {
                List<MenuDish> mds = mdsByMenu.getOrDefault(m.getId(), List.of());
                int dishCount = (int) mds.stream().map(MenuDish::getDishId).filter(Objects::nonNull).distinct().count();
                int reviewedCount = (int) mds.stream().map(MenuDish::getDishId)
                        .filter(Objects::nonNull).filter(myReviewedDishes::contains).distinct().count();
                boolean menuReviewed = myReviewedMenus.contains(m.getId());
                if (!menuReviewed || reviewedCount < dishCount) {
                    pending.add(new MyReviewsVO.PendingMenu(m.getId(), m.getName(),
                            m.getFinishedAt(), dishCount, reviewedCount, menuReviewed));
                }
            }
        }
        return new MyReviewsVO(entries, pending);
    }

    // ===== 纯函数：平均分（TDD 覆盖） =====

    /** 总评星级均分（保留 1 位）。 */
    public BigDecimal averageStar(List<Integer> stars) {
        if (stars == null || stars.isEmpty()) return BigDecimal.ZERO;
        BigDecimal sum = stars.stream().map(BigDecimal::new)
            .reduce(BigDecimal.ZERO, BigDecimal::add);
        return sum.divide(new BigDecimal(stars.size()), 1, RoundingMode.HALF_UP);
    }

    /** 各维度均分：dimensionId -> "x.x"。 */
    public Map<Long, String> averageByDimension(List<? extends DimensionScore> rows) {
        Map<Long, List<Integer>> grouped = rows.stream()
            .filter(r -> r.dimensionId() != null && r.score() != null)
            .collect(Collectors.groupingBy(DimensionScore::dimensionId,
                Collectors.mapping(DimensionScore::score, Collectors.toList())));
        return grouped.entrySet().stream().collect(Collectors.toMap(
            Map.Entry::getKey,
            e -> averageStar(e.getValue()).toPlainString()));
    }

    /** 维度分行接口。 */
    public interface DimensionScore {
        Long dimensionId();
        Integer score();
    }

    // ===== 内部辅助 =====

    /** 当前就餐成员（未登录/未选择返回 null，mine/overview 容错）。 */
    private Long currentMemberId() {
        try {
            return StpUtil.getSession().getLong("currentMemberId");
        } catch (Exception e) {
            return null;
        }
    }

    /** 已按 create_time 倒序的列表 → 最近一条。 */
    private Review latestOne(List<Review> rowsDesc) {
        return rowsDesc == null || rowsDesc.isEmpty() ? null : rowsDesc.get(0);
    }
}
